core/google/api/core/exceptions.py

```diff
@@ -0,0 +1,399 @@
+    pass
+
+
+class _GoogleAPICallErrorMeta(type):
+    """Metaclass for registering GoogleAPICallError subclasses."""
+    def __new__(mcs, name, bases, class_dict):
+        cls = type.__new__(mcs, name, bases, class_dict)
+        if cls.code is not None:
+            _HTTP_CODE_TO_EXCEPTION.setdefault(cls.code, cls)
+        if cls.grpc_status_code is not None:
+            _GRPC_CODE_TO_EXCEPTION.setdefault(cls.grpc_status_code, cls)
+        return cls
+
+
+@six.python_2_unicode_compatible
+@six.add_metaclass(_GoogleAPICallErrorMeta)
+class GoogleAPICallError(GoogleAPIError):
+    """Base class for exceptions raised by calling API methods."""
+
+    code = None
+    """int: The HTTP status code associated with this error.
+
+    See http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html
+    """
+
+    grpc_status_code = None
+    """Optional[grpc.StatusCode]: The gRPC status code associated with this
+    error.
+
+    This may be ``None`` if the exception does not match up to a gRPC error.
+    """
+
+    def __init__(self, message, errors=()):
+        """
+        Args:
+            message (str): The exception message.
+            errors (Sequence[Any]): An optional list of error details.
+        """
+        super(GoogleAPICallError, self).__init__(message)
+        self.message = message
+        self._errors = errors
+
+    def __str__(self):
+        return '{} {}'.format(self.code, self.message)
+
+    @property
+    def errors(self):
+        """Detailed error information.
+
+        Returns:
+            Sequence[Any]: A list of additional error details.
+        """
+        return [copy.deepcopy(error) for error in self._errors]
+
+
+class Redirection(GoogleAPICallError):
+    """Base class for for all redirection (HTTP 3xx) responses."""
+
+
+class MovedPermanently(Redirection):
+    """Exception mapping a ``301 Moved Permanently`` response."""
+    code = http_client.MOVED_PERMANENTLY
+
+
+class NotModified(Redirection):
+    """Exception mapping a ``304 Not Modified`` response."""
+    code = http_client.NOT_MODIFIED
+
+
+class TemporaryRedirect(Redirection):
+    """Exception mapping a ``307 Temporary Redirect`` response."""
+    code = http_client.TEMPORARY_REDIRECT
+
+
+class ResumeIncomplete(Redirection):
+    """Exception mapping a ``308 Resume Incomplete`` response."""
+    # Note: http_client.PERMANENT_REDIRECT is 308, but Google APIs differ
+    # in their use of this status code.
+    code = 308
+
+
+class ClientError(GoogleAPICallError):
+    """Base class for all client error (HTTP 4xx) responses."""
+
+
+class BadRequest(ClientError):
+    """Exception mapping a ``400 Bad Request`` response."""
+    code = http_client.BAD_REQUEST
+
+
+class InvalidArgument(BadRequest):
+    """Exception mapping a :prop:`grpc.StatusCode.INVALID_ARGUMENT` error."""
+    grpc_status_code = (
+        grpc.StatusCode.INVALID_ARGUMENT if grpc is not None else None)
```

Comment:
Agreed. I'm going to leave as-is for now. Changing this will not impact the public interface, so if we find this offensive later we can change without worry.